rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -232,8 +232,12 @@ class Publisher : public PublisherBase
     // interprocess publish, resulting in lower publish-to-subscribe latency.
     // It's not possible to do that with an unique_ptr,
     // as do_intra_process_publish takes the ownership of the message.
+
+    // When durability is set to TransientLocal (i.e. there is a buffer),
+    // inter process publish should always take place to ensure
+    // late joiners receive past data.
     bool inter_process_publish_needed =
-      get_subscription_count() > get_intra_process_subscription_count();
+      get_subscription_count() > get_intra_process_subscription_count() || buffer_;
```

Comment:
We also need this fix in the TypeAdapted signature below, I think.

Reply:
Thanks for catching this!